src/Microsoft.OData.Core/SelectedPropertiesNode.cs

```diff
@@ -336,11 +339,62 @@ internal IEnumerable<IEdmNavigationProperty> GetSelectedNavigationProperties(IEd
             return selectedNavigationProperties.Distinct();
         }
 
+        /// <summary>
+        /// Gets all selected structured properties at the current node.
+        /// </summary>
+        /// <param name="structuredType">The current structured type.</param>
+        /// <returns>The selected structured properties at this node level.</returns>
+        /// <remarks>Result does not include the dynamic properties, which can be retrieved by <code>GetSelectedDynamicProperties</code></remarks>
+        internal IEnumerable<IEdmProperty> GetSelectedProperties(IEdmStructuredType structuredType)
+        {
+            if (this.selectionType == SelectionType.Empty)
+            {
+                return EmptyEdmProperties;
+            }
+
+            // We cannot determine the selected properties without the user model. This means we won't be computing the missing properties.
+            if (structuredType == null)
+            {
+                return EmptyEdmProperties;
+            }
+
+            if (this.selectionType == SelectionType.EntireSubtree || this.hasWildcard)
+            {
+                return structuredType.DeclaredProperties;
+            }
+
+            Debug.Assert(this.selectedProperties != null, "selectedProperties != null");
+
+            // Get declared properties selected, and filter out unrecognized properties.
+            return this.selectedProperties.Select(structuredType.FindProperty).OfType<IEdmProperty>();
+        }
+
+        /// <summary>
+        /// Return names of dynamic properties if current structured node is open type.
+        /// </summary>
+        /// <param name="structuredType">The current structured type.</param>
+        /// <returns>The names of dynamic properties, empty if the current structured type is not open type.</returns>
+        internal IEnumerable<string> GetSelectedDynamicProperties(IEdmStructuredType structuredType)
+        {
+            IEnumerable<string> dynamicProperties = Enumerable.Empty<string>();
+
+            if (structuredType.IsOpen && this.selectedProperties != null && this.selectedProperties.Count > 0)

```

Comment:
>structuredType.IsOpen &&  [](start = 16, length = 25)

remove this check.  the service can always return properties not defined on the type, whether open or not.  if it is in the list of selectedProperties, and not in the payload, we should restore it.

Reply:
ok

---
In reply to: [198068206](https://github.com/OData/odata.net/pull/1106#discussion_r198068206) [](ancestors = 198068206)